### PR TITLE
Testing node sockets instead of jot

### PR DESCRIPTION
**Pull request by hejule**
- Changes `json-over-tcp` to node's own Socket
- Fixes movement & stamping logic
- Fix tests

### CODE REVIEW

File: src/util/parser.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+export const tryParseObject = (data: string | Buffer): Record<string, any> => {
+  try {
+    if (Buffer.isBuffer(data)) {
+      return JSON.parse(data.toString())
+    }
+    const dataString = data.toString()
+    const startIndex = dataString.indexOf('{')
+    const endIndex = dataString.lastIndexOf('}') + 2
+    const messageString = dataString.substr(startIndex, endIndex - startIndex)
```
Comment (mikkoj) [inline src/util/parser.ts]: Why is this required? Wondering how performant this is if there are loads of messages?

File: src/util/parser.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+export const tryParseObject = (data: string | Buffer): Record<string, any> => {
```
Comment (mikkoj) [inline src/util/parser.ts]: `Record<string, any>` is used for "anything json" here?

File: src/util/parser.ts (hunk 1 above)
Comment (hejule) [reply src/util/parser.ts]: Yep, just hacked this together to get forward but this data receiving is the thing I meant when I mentioned that it would be nice to have a library handle it.
Receiving partial messages or such isn't handled at all (not sure if that can really happen but theoretically it could)

File: src/util/parser.ts (hunk 2 above)
Comment (hejule) [reply src/util/parser.ts]: Hm, yeah I thought we'd need to type the responses but actually we can replace that with `Message` :D

File: src/server.ts [hunk 3]
```diff
@@ -43,7 +45,30 @@ const clientConnectionHandler = (socket: Socket): void => {
   })
 }
 
+const shutDown = (eventHistory: Array<HistoryEvent>): void => {
```
Comment (mikkoj) [inline src/server.ts]: Could this be async so we could write:
```TypeScript
try {
  await fs.writeFile(...)
} catch(err) {
  Log.error(...)
}
```
🤓

File: src/server.ts [hunk 4]
```diff
@@ -1,7 +1,9 @@
 import Pummiralli from './domain/pummiralli'
+import { HistoryEvent } from './domain/messages'
 import { createServer, Socket, AddressInfo } from 'net'
 import Log from './util/log'
 import { tryParseObject } from './util/parser'
+import fs from 'fs'
 
 // import Ajv from 'ajv'
 // import schema from './messages/schema.json'
```
Comment (mikkoj) [inline src/server.ts]: Should we remove all these schema thingies for now?

File: src/domain/bot.ts [hunk 5]
```diff
@@ -43,4 +52,14 @@ export default class Bot {
       }`,
     )
   }
+  handleFinish(map: Map): void {
+    const { x, y } = this.position
+    const closeEnoughToCheckpoint = map.closeEnoughToCheckpoint(this.position)
+    console.log(
```
Comment (mikkoj) [inline src/domain/bot.ts]: Use Log.something here to avoid `console`? 😉

File: src/domain/messages.ts [hunk 6]
```diff
@@ -99,6 +105,7 @@ export type Message =
   | GameEndMessage
   | PlayerPositionsMessage
   | StampMessage
+  | FinishMessage
   | MapMessage
   | InvalidMessage
```
Comment (mikkoj) [inline src/domain/messages.ts]: nit: Was thining whether it would make sense to separate these to:
`MessageByServer` and `MessageByClient` to avoid mistakes.

```TypeScript
export type MessageByServer =
  | ErrorMessage
  | MapMessage
  | GameStartMessage
  | GameEndMessage
  | PlayerPositionsMessage
  | JoinMessage

export type MessageByClient = JoinMessage | MoveMessage | StampMessage | FinishMessage

export type Message = MessageByServer | MessageByClient
```

File: src/server.ts [hunk 7]
```diff
@@ -43,7 +45,30 @@ const clientConnectionHandler = (socket: Socket): void => {
   })
 }
 
+const shutDown = (eventHistory: Array<HistoryEvent>): void => {
+  const resultsDir = './.results'
```
Comment (mikkoj) [inline src/server.ts]: Should we add this in `.gitignore`? 🤔

File: src/domain/messages.ts (hunk 6 above)
Comment (hejule) [reply src/domain/messages.ts]: 👍

File: src/server.ts (hunk 4 above)
Comment (hejule) [reply src/server.ts]: 👍

File: src/server.ts (hunk 7 above)
Comment (hejule) [reply src/server.ts]: added

File: src/server.ts (hunk 3 above)
Comment (hejule) [reply src/server.ts]: changed 👍
